### PR TITLE
stream: move throttle.rs

### PR DESCRIPTION
**Pull request by blasrodri**
Ref: #2727

### CODE REVIEW

**Comment by blasrodri** [comment]
> The throttle type needs to go in its own file.

And how is it related to `StreamExt `?

File: tokio/src/stream/mod.rs [hunk 1]
```diff
@@ -835,3 +878,74 @@ fn merge_size_hints(
...
+        duration: Duration,
+
+        // Set to true when `delay` has returned ready, but `stream` hasn't.
+        has_delayed: bool,
+
+        // The stream to throttle
+        #[pin]
+        stream: T,
+    }
+}
```
Comment (Darksonn) [inline tokio/src/stream/mod.rs:900]: What I mean is that _this_ should not be in `tokio/src/stream/mod.rs`, but rather something like `tokio/src/stream/throttle.rs`.

File: tokio/src/stream/throttle.rs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 //! Slow down a stream by enforcing a delay between items.
 
 use crate::stream::Stream;
-use crate::time::{Delay, Duration, Instant};
+use crate::stream::{Delay, Duration, Instant};
```
Comment (Darksonn) [inline tokio/src/stream/throttle.rs:4]: Something seems wrong here. These types should be in the `time` module.

**Comment by blasrodri** [comment]
A method that internally calls that public function?

**Comment by Darksonn** [comment]
I was thinking of making the free-standing function private, and providing access only through `StreamExt`, just like with [`StreamExt::timeout`][1].

[1]: https://docs.rs/tokio/0.2/tokio/stream/trait.StreamExt.html#method.timeout

File: tokio/src/stream/mod.rs [hunk 3]
```diff
@@ -819,6 +821,33 @@ pub trait StreamExt: Stream {
     {
         Timeout::new(self, duration)
     }
+    /// Slows down a stream by enforcing a delay between items.
+    /// They will be produced not more often than the specified interval.
```
Comment (Darksonn) [inline tokio/src/stream/mod.rs:825]: I know you just copy-pasted this, but I think the sentence is awkward.

**Review by Darksonn — APPROVED** [review]
I think this is good. Feel free to do the change below, but it is not needed for a merge.

File: tokio/src/stream/mod.rs (hunk 3 above)
Comment (blasrodri) [reply tokio/src/stream/mod.rs:825]: Perhaps remove `/// They will be produced not more often than the specified interval.` would improve?

Comment (Darksonn) [reply tokio/src/stream/mod.rs:825]: I guess it makes sense.

**Review by Darksonn — APPROVED** [review]

**Comment by webern** [comment]
I have some code that depends on `tokio::time::throttle`. Is the function still available as part of tokio's public API, or did this remove it from the public API?

**Comment by Darksonn** [comment]
It was moved to tokio-stream as [`StreamExt::throttle`][1].

[1]: https://docs.rs/tokio-stream/0.1/tokio_stream/trait.StreamExt.html#method.throttle

**Comment by webern** [comment]
Thank you!

**Pull request merged**